### PR TITLE
 i#2390 ind br: Emit NOPs instead of movk reg, #0. 

### PR DESCRIPTION
**Pull request by fhahn**
Instead of creating unnecessary movk instructions when creating
exit stubs, we insert nops instead just before the branch. Ideally we
would just have a variable size for the exit stub fragments, but at
different places DynamoRIO expects them the be of a fixed size and the
last instruction to be a branch.

Issue #2390

### CODE REVIEW

**Comment by derekbruening** [comment]
Right, making exit stubs variable-sized would take extra work.  Separate stubs are allocated from a heap specialized for constant-size allocations.  I would argue it is not worth the effort to make them variable when stubs are allocated separately, as too-large stubs do not affect the fastpath when fragments are linked.

Re: the jump being last: the assumption is in `link_indirect_exit_arch`.  What is your proposal for finding the jump if it's not last?

Also xref (and `link_indirect_exit_arch` for x86 should show up here: its hardcoded `5` should instead be `JMP_LONG_LENGTH'`):
```
$ git grep ' \- JMP_LONG_LENGTH'
core/arch/emit_utils_shared.c:    jmp = stub + stub_size - JMP_LONG_LENGTH;
core/arch/emit_utils_shared.c:     *   trace_head_return_prefix - JMP_LONG_LENGTH - 4
core/arch/emit_utils_shared.c:        (info->trace_head_return_prefix - JMP_LONG_LENGTH - sizeof(info));
core/arch/x86/emit_utils.c:        return (stub + STUB_COARSE_DIRECT_SIZE64 - JMP_LONG_LENGTH);
core/arch/x86/emit_utils.c:    return (stub + STUB_COARSE_DIRECT_SIZE32 - JMP_LONG_LENGTH);
```
Those are all for coarse stubs which are not yet supported for aarch.

**Comment by fhahn** [comment]
> Re: the jump being last: the assumption is in link_indirect_exit_arch. What is your proposal for finding the jump if it's not last?

If we insert the NOPs after the branch, we would at most have to skip 3 nops to find the branch. I just tried it and that mostly works. Only the signal test cases are failing with that, because it fails to re-create the app state. I try to do some more digging.

**Comment by derekbruening** [comment]
> > Re: the jump being last: the assumption is in link_indirect_exit_arch. What is your proposal for finding the jump if it's not last?
>
> If we insert the NOPs after the branch, we would at most have to skip 3 nops to find the branch. I just tried it and that mostly works. Only the signal test cases are failing with that, because it fails to re-create the app state. I try to do some more digging.

You mean, you're going to decode the instructions in reverse until you see a jump?  Presumably via a direct pattern-match to avoid the cost of full decoding which we want to avoid.

**Comment by fhahn** [comment]
> You mean, you're going to decode the instructions in reverse until you see a jump? Presumably via a direct pattern-match to avoid the cost of full decoding which we want to avoid.

Yep, we just need to check the raw encoding. I've pushed an update doing that, what do you think?

Currently it still fails most linux.signal* tests, that's what I have to investigate further.

File: core/arch/aarch64/emit_utils.c [hunk 1]
```diff
@@ -127,12 +137,22 @@ insert_exit_stub_other_flags(dcontext_t *dcontext, fragment_t *f,
...
         pc = insert_mov_imm(pc, DR_REG_X0, (ptr_int_t)l);
+        num_nops_needed = 4 - (pc - (uint *) stub_pc - 1);
         /* ldr x1, [x(stolen), #(offs)] */
         *pc++ = (0xf9400000 | 1 | (dr_reg_stolen - DR_REG_X0) << 5 |
                  get_ibl_entry_tls_offs(dcontext, exit_target) >> 3 << 10);
+
         /* br x1 */
         *pc++ = 0xd61f0000 | 1 << 5;
     }
+        /* Fill up with NOPs, depending on how many instructions we needed to move
```
Comment (derekbruening) [inline core/arch/aarch64/emit_utils.c]: Indentation looks off for this set of lines

File: core/arch/aarch64/emit_utils.c [hunk 2]
```diff
@@ -127,12 +137,22 @@ insert_exit_stub_other_flags(dcontext_t *dcontext, fragment_t *f,
...
+         * the immediate in a registers. Ideally we would skip adding NOPs, but
+         * the lots of places expect the stub size to be fixed.
```
Comment (derekbruening) [inline core/arch/aarch64/emit_utils.c]: s/the lots/lots/

File: core/arch/aarch64/emit_utils.c [hunk 3]
```diff
@@ -250,9 +270,14 @@ link_indirect_exit_arch(dcontext_t *dcontext, fragment_t *f,
...
         exit_target = get_linked_entry(dcontext, target_tag);
 
-    pc = (uint *)(stub_pc + exit_stub_size(dcontext, target_tag, f->flags) -
-                  (2 * AARCH64_INSTR_SIZE));
+    /* Set pc to the last instruction in the stub. */
+    pc = (uint *)(stub_pc + exit_stub_size(dcontext, target_tag, f->flags) - AARCH64_INSTR_SIZE);
+
+    /* Skip NOP instructions backwards. First non-NOP instruction must be the branch. */
+    while (*pc == NOP_INST)
+        pc--;
```
Comment (derekbruening) [inline core/arch/aarch64/emit_utils.c]: Can we have a debug-build check that we are now at a branch?

File: core/arch/aarch64/emit_utils.c [hunk 4]
```diff
@@ -250,9 +270,14 @@ link_indirect_exit_arch(dcontext_t *dcontext, fragment_t *f,
...
 
+    pc -= 1;
     /* ldr x1, [x(stolen), #(offs)] */
     pc[0] = (0xf9400000 | 1 | (dr_reg_stolen - DR_REG_X0) << 5 |
              get_ibl_entry_tls_offs(dcontext, exit_target) >> 3 << 10);
```
Comment (derekbruening) [inline core/arch/aarch64/emit_utils.c]: Probably unlink_indirect_exit also needs to skip the final nops

File: core/arch/aarch64/emit_utils.c [hunk 5]
```diff
@@ -253,6 +254,16 @@ exit_cti_disp_pc(cache_pc branch_pc)
     return NULL;
 }
 
+/* Skips NOP instructions backwards until the first non-NOP instruction is found. */
+static uint *get_stub_branch(uint *pc) {
```
Comment (derekbruening) [inline core/arch/aarch64/emit_utils.c]: style: { on own line

File: core/arch/aarch64/emit_utils.c [hunk 6]
```diff
@@ -250,17 +281,17 @@ link_indirect_exit_arch(dcontext_t *dcontext, fragment_t *f,
...
     pc = (uint *)(stub_pc + exit_stub_size(dcontext, target_tag, f->flags) -
-                  (2 * AARCH64_INSTR_SIZE));
+                  AARCH64_INSTR_SIZE);
 
+    pc = get_stub_branch(pc) - 1;
     /* ldr x1, [x(stolen), #(offs)] */
-    pc[0] = (0xf9400000 | 1 | (dr_reg_stolen - DR_REG_X0) << 5 |
-             get_ibl_entry_tls_offs(dcontext, exit_target) >> 3 << 10);
-    /* br x1 */
-    pc[1] = 0xd61f0000 | 1 << 5;
```
Comment (derekbruening) [inline core/arch/aarch64/emit_utils.c]: Was this a redundant write before?

Comment (fhahn) [reply core/arch/aarch64/emit_utils.c]: Yes looks like it. The assertion in get_stub_branch should catch the case where it isn't.

**Comment by fhahn** [comment]
Thank you very much for pointing me in the right directions :)

**Pull request merged**